cluster-autoscaler/proposals/plugable-provider-grpc.md

```diff
@@ -0,0 +1,604 @@
+
+  k8s.io.apimachinery.pkg.apis.meta.v1.Time endTime = 3;
+}
+
+message PricingNodePriceResponse {
+ // Price of the theoretical minimum price of running a pod for a given period
+ float64 price = 1;
+}
+```
+
+### PricingPodPrice
+
+PodPrice handles an operation that returns a theoretical minimum price of running a pod for a given period of time on a perfectly matching machine.
+
+```protobuf
+message PricingPodPriceRequest {
+  k8s.io.api.core.v1.Pod pod = 1;,
+
+  k8s.io.apimachinery.pkg.apis.meta.v1.Time startTime = 2;
+
+  k8s.io.apimachinery.pkg.apis.meta.v1.Time endTime = 3;
+}
+
+message PricingPodPriceResponse {
+ // Price of the theoretical minimum price of running a pod for a given period
+ float64 price = 1;
+}
+```
+
+### NewNodeGroup
+
+NewNodeGroup builds a theoretical node group based on the node definition provided. The node group is not automatically created on the cloud provider side. This action returns the created node group.
+
+```protobuf
+message NewNodeGroupRequest {
+  // Machine type for the node group.
+  string machineType = 1;
+
+  // Labels of the node group.
+  map<string,string> labels = 2;
+
+  // System Labels of the node group.
+  map<string, string> systemLabels = 3;
+
+  // Taints of the node group
+  repeated k8s.io.api.core.v1.Taint taints = 4;
+
+  // ExtraResources of the node group
+  map<string, k8s.io.apimachinery.pkg.api.resource.Quantity extraResources = 5;
+}
+
+message NewNodeGroupResponse {
+  // NodeGroup created by the cloud provider.
+ NodeGroup nodeGroup = 1;
+}
+```
+
+### GetResourceLimiter
+
+GetResourceLimiter holds a struct containing limits (max, min) for resources (cores, memory etc.).
+
+```protobuf
+message GetResourceLimiterRequest {
+  // Intentionally empty.
+}
+
+message GetResourceLimiterResponse {
+ // Contains the minimum limits for resources (cores, memory etc.).
+ map<string, int64> minLimits = 1;
+
+ // Contains the maximum limits for resources (cores, memory etc.).
+ map<string, int64> maxLimits = 2;
+}
+```
+
+### Refresh
+
+Refresh is called before every main loop and can be used to dynamically update cloud provider state.
+In particular the list of node groups returned by NodeGroups can change as a result of this action.
+
+```protobuf
+message RefreshRequest {
+  // Intentionally empty.
+}
+
+message RefreshResponse {
+// Intentionally empty.
+}
+```
+
+### Cleanup
+
+Cleanup cleans up open resources before the cloud provider is destroyed, i.e. go routines etc.
+
+```protobuf
+message CleanupRequest {
+  // Intentionally empty.
+}
+
+message CleanupResponse {
+// Intentionally empty.
+}
+```
+
+### Name
+
+Name stores the name of the cloud provider.
+
+```protobuf
+message NameRequest {
+  // Intentionally empty.
+}
+
+message NameResponse {
+  // Name of the node group
+ string name = 1;
+}
+```
+
+### NodeGroup
+
+```protobuf
+message NodeGroup {
+  // Name of the node group on the cloud provider
+  string name = 1;
```

Comment:
Delete this? It seems to duplicate id below.

Reply:
Yeah, it is a duplicate of the id, removing it!.